tests/e2e-evm/test/abi_basic.test.ts

```diff
@@ -0,0 +1,162 @@
+import hre from "hardhat";
+import type { ArtifactsMap } from "hardhat/types/artifacts";
+import type { PublicClient, WalletClient, ContractName } from "@nomicfoundation/hardhat-viem/types";
+import { expect } from "chai";
+import { Address, toFunctionSelector, toFunctionSignature, concat } from "viem";
+import { getAbiFallbackFunction, getAbiReceiveFunction } from "./helpers/abi";
+import { whaleAddress } from "./addresses";
+
+interface ContractTestCase {
+  interface: keyof ArtifactsMap;
+  mock: ContractName<keyof ArtifactsMap>;
+}
+
+// ABI_BasicTests assert ethereum + solidity transaction ABI interactions perform as expected.
+describe("ABI_BasicTests", function () {
+  const testCases = [
+    // Test function modifiers without receive & fallback
+    { interface: "NoopNoReceiveNoFallback", mock: "NoopNoReceiveNoFallbackMock" },
+
+    // Test receive + fallback scenarios
+    { interface: "NoopReceiveNoFallback", mock: "NoopReceiveNoFallbackMock" },
+    { interface: "NoopReceivePayableFallback", mock: "NoopReceivePayableFallbackMock" },
+    { interface: "NoopReceiveNonpayableFallback", mock: "NoopReceiveNonpayableFallbackMock" },
+
+    // Test no receive + fallback scenarios
+    { interface: "NoopNoReceivePayableFallback", mock: "NoopNoReceivePayableFallbackMock" },
+    { interface: "NoopNoReceiveNonpayableFallback", mock: "NoopNoReceiveNonpayableFallbackMock" },
+  ] as ContractTestCase[];
```

Comment:
nit: i'd prefer to declare type this like `const testCases: ContractTestCase[] =  ...` rather than cast with `as`. 
it's a subtle difference, but i think of it like (declaring) "hey, compiler, these are supposed to be `ContractTestCase`s, plz help" vs (casting) "hey, compiler, these ARE `ContractTestCase`"

Reply:
I 100% agree and it was the first thing I tried, but typescript was unable to correctly infer contract name type for the mock property with an `as ContractTestCase[]`, so didn't think it was worth duplicating.

Maybe there is another way to hint at the type?  Explicitly typing the strings in mock also worked, but it has to be repeated for every element.

Reply:
types are weird here yeah since they're generated in the artifacts? I tried some wonky thing like this which works for this section but causes some `"Unsafe assignment of an error typed value."` issues below when using `hre.artifacts.readArtifactSync(tc.interface).abi;`

```typescript
interface ContractTestCase<T extends keyof ArtifactsMap> {
  interface: T;
  mock: ArtifactsMap[T]["contractName"];
}

// ABI_BasicTests assert ethereum + solidity transaction ABI interactions perform as expected.
describe("ABI_BasicTests", function () {
  const testCases: ContractTestCase<keyof ArtifactsMap>[] = [
    // Test function modifiers without receive & fallback
    { interface: "NoopNoReceiveNoFallback", mock: "NoopNoReceiveNoFallbackMock" },

    // Test receive + fallback scenarios
    { interface: "NoopReceiveNoFallback", mock: "NoopReceiveNoFallbackMock" },
    { interface: "NoopReceivePayableFallback", mock: "NoopReceivePayableFallbackMock" },
    { interface: "NoopReceiveNonpayableFallback", mock: "NoopReceiveNonpayableFallbackMock" },

    // Test no receive + fallback scenarios
    { interface: "NoopNoReceivePayableFallback", mock: "NoopNoReceivePayableFallbackMock" },
    { interface: "NoopNoReceiveNonpayableFallback", mock: "NoopNoReceiveNonpayableFallbackMock" },
  ];
...
```

Reply:
heh, potentially hacky workaround that skirts the `ContractName` type troubles:
```
interface ContractTestCase {
  interface: keyof ArtifactsMap;
  mock: `${keyof ArtifactsMap}Mock`;
}
```
but not a major concern lol